File: ios/SnapSell/UI/Screens/Confirm/ConfirmViewModel.swift
import Foundation
import Observation

struct ConfirmCard: Identifiable {
    var itemId: String
    var photoFile: String
    var name = ""
    var brand = ""
    var model = ""
    var category = ""
    var condition: Condition = .good
    var notes = ""
    var attributes: [String] = []
    var searchQuery = ""
    var confidence = 0.0
    var identifying = false
    var identifyError: String?
    var pricing = false
    var priceError: String?
    var quote: PriceQuote?
    /// True after the user confirmed the identity at least once.
    var confirmed = false

    var id: String { itemId }
    var identified: Bool { !name.isBlank }
    var busy: Bool { identifying || pricing }
}

/// Identifies each placeholder item (created by Review), lets the user edit
/// identity + condition, and prices on confirm. Editing name/model/condition
/// after a price exists re-runs /price after a short debounce.
@MainActor
@Observable
final class ConfirmViewModel {
    static let repriceDebounceMs: UInt64 = 1200

    private(set) var loading = true
    private(set) var cards: [ConfirmCard] = []
    private(set) var error: String?

    var allPriced: Bool { !cards.isEmpty && cards.allSatisfy { $0.quote != nil } }
    var anyBusy: Bool { cards.contains { $0.busy } }

    @ObservationIgnored private let app: AppContainer
    /// Backend calls are heavy (vision LLM); keep at most two in flight.
    @ObservationIgnored private let identifyLimiter = AsyncSemaphore(limit: 2)
    @ObservationIgnored private var repriceTasks: [String: Task<Void, Never>] = [:]

    init(app: AppContainer, itemIds: [String]) {
        self.app = app
        let items = app.inventory.items(itemIds)
        cards = items.map(Self.card)
        loading = false
        error = items.isEmpty ? "Nothing to confirm." : nil
        for item in items where item.name.isBlank { identify(item.id) }
    }

    private static func card(_ item: Item) -> ConfirmCard {
        ConfirmCard(
            itemId: item.id, photoFile: item.photoFile, name: item.name, brand: item.brand.orEmpty, model: item.model.orEmpty,
            category: item.category, condition: item.condition, notes: item.notes.orEmpty, attributes: item.attributes,
            searchQuery: item.searchQuery, confidence: item.confidence, quote: item.quote, confirmed: item.quote != nil
        )
    }

    private func update(_ id: String, _ transform: (inout ConfirmCard) -> Void) {
        guard let i = cards.firstIndex(where: { $0.itemId == id }) else { return }
        transform(&cards[i])
    }

    func card(_ id: String) -> ConfirmCard? { cards.first { $0.itemId == id } }

    func identify(_ id: String) {
        update(id) { $0.identifying = true; $0.identifyError = nil }
        Task {
            guard let card = card(id) else { return }
            await identifyLimiter.acquire()
            let result = await app.identify.identify(imageURL: app.photoStore.url(card.photoFile), hint: nil)
            identifyLimiter.release()
            switch result {
            case .success(let r):
                if var existing = app.inventory.item(id) {
                    existing.name = r.item.name
                    existing.brand = r.item.brand
                    existing.model = r.item.model
                    existing.category = r.item.category
                    existing.condition = r.item.condition
                    existing.attributes = r.item.attributes
                    existing.searchQuery = r.item.searchQuery
                    existing.confidence = r.item.confidence
                    existing.notes = r.item.notes
                    existing.listingTitle = r.listingText.title
                    existing.listingDescription = r.listingText.description
                    try? app.inventory.saveItem(existing)
                }
                update(id) {
                    $0.identifying = false
                    $0.name = r.item.name; $0.brand = r.item.brand.orEmpty; $0.model = r.item.model.orEmpty
                    $0.category = r.item.category; $0.condition = r.item.condition; $0.notes = r.item.notes.orEmpty
                    $0.attributes = r.item.attributes; $0.searchQuery = r.item.searchQuery; $0.confidence = r.item.confidence
                }
            case .failure(let e):
                update(id) { $0.identifying = false; $0.identifyError = e.message }
            }
        }
    }

    func editName(_ id: String, _ v: String) { edit(id, reprice: true) { $0.name = v } }
    func editBrand(_ id: String, _ v: String) { edit(id, reprice: false) { $0.brand = v } }
    func editModel(_ id: String, _ v: String) { edit(id, reprice: true) { $0.model = v } }
    func editNotes(_ id: String, _ v: String) { edit(id, reprice: false) { $0.notes = v } }
    func editCondition(_ id: String, _ v: Condition) { edit(id, reprice: true) { $0.condition = v } }

    private func edit(_ id: String, reprice: Bool, _ transform: (inout ConfirmCard) -> Void) {
        update(id, transform)
        guard let card = card(id) else { return }
        if reprice, card.quote != nil { scheduleReprice(id) }
    }

    private func scheduleReprice(_ id: String) {
        repriceTasks[id]?.cancel()
        repriceTasks[id] = Task {
            try? await Task.sleep(nanoseconds: Self.repriceDebounceMs * 1_000_000)
            guard !Task.isCancelled else { return }
            confirm(id)
        }
    }

    /// Persist edits and fetch a price.
    func confirm(_ id: String) {
        guard let card = card(id), card.identified, !card.pricing else { return }
        repriceTasks.removeValue(forKey: id)?.cancel()
        update(id) { $0.pricing = true; $0.priceError = nil; $0.confirmed = true }
        Task {
            guard var updated = app.inventory.item(id) else { return }
            let name = card.name.trimmingCharacters(in: .whitespacesAndNewlines)
            let existingName = updated.name
            updated.name = name
            updated.brand = card.brand.trimmingCharacters(in: .whitespacesAndNewlines).nonBlank
            updated.model = card.model.trimmingCharacters(in: .whitespacesAndNewlines).nonBlank
            updated.condition = card.condition
            updated.notes = card.notes.trimmingCharacters(in: .whitespacesAndNewlines).nonBlank
            updated.category = card.category.isBlank ? "Uncategorized" : card.category
            // Keep the model's query unless the user changed the name; then the name is the better query.
            updated.searchQuery = name != existingName ? name : (updated.searchQuery.isBlank ? name : updated.searchQuery)
            try? app.inventory.saveItem(updated)
            let factor = app.settings.localSaleFactor
            switch await app.pricing.price(item: updated.toDto(), localSaleFactor: factor) {
            case .success(let q):
                try? app.inventory.saveQuote(itemId: id, quote: q)
                update(id) { $0.pricing = false; $0.quote = q; $0.searchQuery = updated.searchQuery }
            case .failure(let e):
                update(id) { $0.pricing = false; $0.priceError = e.message }
            }
        }
    }

    func confirmAll() {
        for card in cards where card.identified && !card.pricing { confirm(card.itemId) }
    }

    func remove(_ id: String) {
        repriceTasks.removeValue(forKey: id)?.cancel()
        try? app.inventory.deleteItem(id)
        cards.removeAll { $0.itemId == id }
    }
}

/// Counting semaphore for async code.
actor AsyncSemaphore {
    private let limit: Int
    private var inUse = 0
    private var waiters: [CheckedContinuation<Void, Never>] = []

    init(limit: Int) { self.limit = limit }

    func acquire() async {
        if inUse < limit { inUse += 1; return }
        await withCheckedContinuation { waiters.append($0) }
    }

    nonisolated func release() { Task { await self.releaseInner() } }

    private func releaseInner() {
        if let next = waiters.first { waiters.removeFirst(); next.resume() } else { inUse -= 1 }
    }
}

File: android/app/src/main/java/com/alexivanov/snapsell/ui/confirm/ConfirmViewModel.kt
package com.alexivanov.snapsell.ui.confirm

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.alexivanov.snapsell.AppContainer
import com.alexivanov.snapsell.core.AppResult
import com.alexivanov.snapsell.data.local.ItemEntity
import com.alexivanov.snapsell.data.local.toDto
import com.alexivanov.snapsell.data.remote.dto.PriceQuote
import com.alexivanov.snapsell.domain.Condition
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import java.io.File

data class ConfirmCard(
    val itemId: String,
    val photoPath: String,
    val name: String = "",
    val brand: String = "",
    val model: String = "",
    val category: String = "",
    val condition: Condition = Condition.GOOD,
    val notes: String = "",
    val attributes: List<String> = emptyList(),
    val searchQuery: String = "",
    val confidence: Double = 0.0,
    val identifying: Boolean = false,
    val identifyError: String? = null,
    val pricing: Boolean = false,
    val priceError: String? = null,
    val quote: PriceQuote? = null,
    /** True after the user confirmed the identity at least once. */
    val confirmed: Boolean = false,
) {
    val identified: Boolean get() = name.isNotBlank()
    val busy: Boolean get() = identifying || pricing
}

data class ConfirmUiState(
    val loading: Boolean = true,
    val cards: List<ConfirmCard> = emptyList(),
    val error: String? = null,
) {
    val allPriced: Boolean get() = cards.isNotEmpty() && cards.all { it.quote != null }
    val anyBusy: Boolean get() = cards.any { it.busy }
}

/**
 * Identifies each placeholder item (created by Review), lets the user edit
 * identity + condition, and prices on confirm. Editing name/model/condition
 * after a price exists re-runs /price after a short debounce, so the quote
 * never silently refers to a different item.
 */
class ConfirmViewModel(private val container: AppContainer, private val itemIds: List<String>) : ViewModel() {
    private val _state = MutableStateFlow(ConfirmUiState())
    val state: StateFlow<ConfirmUiState> = _state

    /** Backend calls are heavy (vision LLM); keep at most two in flight. */
    private val identifyLimiter = Semaphore(2)
    private val repriceJobs = HashMap<String, Job>()

    init {
        viewModelScope.launch {
            val items = container.inventory.getItems(itemIds)
            _state.update {
                it.copy(loading = false, cards = items.map(::cardFrom), error = if (items.isEmpty()) "Nothing to confirm." else null)
            }
            items.filter { it.name.isBlank() }.forEach { identify(it.id) }
        }
    }

    private fun cardFrom(item: ItemEntity) = ConfirmCard(
        itemId = item.id,
        photoPath = item.photoPath,
        name = item.name,
        brand = item.brand.orEmpty(),
        model = item.model.orEmpty(),
        category = item.category,
        condition = item.condition,
        notes = item.notes.orEmpty(),
        attributes = item.attributes,
        searchQuery = item.searchQuery,
        confidence = item.confidence,
        quote = item.quote,
        confirmed = item.quote != null,
    )

    private fun updateCard(id: String, transform: (ConfirmCard) -> ConfirmCard) {
        _state.update { s -> s.copy(cards = s.cards.map { if (it.itemId == id) transform(it) else it }) }
    }

    fun identify(id: String) {
        updateCard(id) { it.copy(identifying = true, identifyError = null) }
        viewModelScope.launch {
            val card = _state.value.cards.firstOrNull { it.itemId == id } ?: return@launch
            val result = identifyLimiter.withPermit { container.identify.identify(File(card.photoPath)) }
            when (result) {
                is AppResult.Success -> {
                    val r = result.value
                    val existing = container.inventory.getItem(id)
                    if (existing != null) {
                        container.inventory.saveItem(
                            existing.copy(
                                name = r.item.name,
                                brand = r.item.brand,
                                model = r.item.model,
                                category = r.item.category,
                                condition = r.item.condition,
                                attributes = r.item.attributes,
                                searchQuery = r.item.searchQuery,
                                confidence = r.item.confidence,
                                notes = r.item.notes,
                                listingTitle = r.listingText.title,
                                listingDescription = r.listingText.description,
                            ),
                        )
                    }
                    updateCard(id) {
                        it.copy(
                            identifying = false,
                            name = r.item.name,
                            brand = r.item.brand.orEmpty(),
                            model = r.item.model.orEmpty(),
                            category = r.item.category,
                            condition = r.item.condition,
                            notes = r.item.notes.orEmpty(),
                            attributes = r.item.attributes,
                            searchQuery = r.item.searchQuery,
                            confidence = r.item.confidence,
                        )
                    }
                }
                is AppResult.Failure -> updateCard(id) { it.copy(identifying = false, identifyError = result.message) }
            }
        }
    }

    fun editName(id: String, v: String) = edit(id, reprice = true) { it.copy(name = v) }
    fun editBrand(id: String, v: String) = edit(id, reprice = false) { it.copy(brand = v) }
    fun editModel(id: String, v: String) = edit(id, reprice = true) { it.copy(model = v) }
    fun editNotes(id: String, v: String) = edit(id, reprice = false) { it.copy(notes = v) }
    fun editCondition(id: String, v: Condition) = edit(id, reprice = true) { it.copy(condition = v) }

    private fun edit(id: String, reprice: Boolean, transform: (ConfirmCard) -> ConfirmCard) {
        updateCard(id, transform)
        val card = _state.value.cards.firstOrNull { it.itemId == id } ?: return
        if (reprice && card.quote != null) scheduleReprice(id)
    }

    private fun scheduleReprice(id: String) {
        repriceJobs[id]?.cancel()
        repriceJobs[id] = viewModelScope.launch {
            delay(REPRICE_DEBOUNCE_MS)
            confirm(id)
        }
    }

    /** Persist edits and fetch a price. */
    fun confirm(id: String) {
        val card = _state.value.cards.firstOrNull { it.itemId == id } ?: return
        if (!card.identified || card.pricing) return
        repriceJobs.remove(id)?.cancel()
        updateCard(id) { it.copy(pricing = true, priceError = null, confirmed = true) }
        viewModelScope.launch {
            val existing = container.inventory.getItem(id) ?: return@launch
            val updated = existing.copy(
                name = card.name.trim(),
                brand = card.brand.trim().ifBlank { null },
                model = card.model.trim().ifBlank { null },
                condition = card.condition,
                notes = card.notes.trim().ifBlank { null },
                category = card.category.ifBlank { "Uncategorized" },
                // Keep the model's query unless the user changed the name; then the name is the better query.
                searchQuery = if (card.name.trim() != existing.name) card.name.trim() else existing.searchQuery.ifBlank { card.name.trim() },
            )
            container.inventory.saveItem(updated)
            val factor = container.settings.currentLocalSaleFactor()
            when (val r = container.pricing.price(updated.toDto(), factor)) {
                is AppResult.Success -> {
                    container.inventory.saveQuote(id, r.value)
                    updateCard(id) { it.copy(pricing = false, quote = r.value, searchQuery = updated.searchQuery) }
                }
                is AppResult.Failure -> updateCard(id) { it.copy(pricing = false, priceError = r.message) }
            }
        }
    }

    fun confirmAll() {
        _state.value.cards.filter { it.identified && !it.pricing }.forEach { confirm(it.itemId) }
    }

    fun remove(id: String) {
        repriceJobs.remove(id)?.cancel()
        viewModelScope.launch {
            container.inventory.deleteItem(id)
            _state.update { s -> s.copy(cards = s.cards.filterNot { it.itemId == id }) }
        }
    }

    companion object {
        const val REPRICE_DEBOUNCE_MS = 1200L
    }
}
